Add binary operators generator

matedabis opened the pull request:
All binary operators (+ - / % * for arithmetic and ^ | & << >> >>> for bitwise operations) are included and working.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+def get_arguments():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--op", help = "give the operators to be tested", nargs = '+',
+    default = ["+", "-", "/", "*", "%", "^"], choices = ["+", "-", "/", "*", "%", "^"])
```
Please separate ` ["+", "-", "/", "*", "%", "^"] ` into a variable instead of duplicate the declaration.

matedabis replied:
Created a variable.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+
+def get_arguments():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--op", help = "give the operators to be tested", nargs = '+',
```
`help = "List of binary operator(s) from an expression is being generated"`

matedabis replied:
Changed description. Maybe we will create a function for the parser, because mine and @repasics's parser only differs in the possible arguments of `--op`, and it may be useful for other test generator scripts.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--op", help = "give the operators to be tested", nargs = '+',
+    default = ["+", "-", "/", "*", "%", "^"], choices = ["+", "-", "/", "*", "%", "^"])
+    parser.add_argument("--test-count", help = "give the number of tests to run", nargs = 1, default = [500],
```
` help = "Number of tests to generate" `

matedabis replied:
Changed description.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+    default = ["+", "-", "/", "*", "%", "^"], choices = ["+", "-", "/", "*", "%", "^"])
+    parser.add_argument("--test-count", help = "give the number of tests to run", nargs = 1, default = [500],
+    type = int)
+    parser.add_argument("--output", help = "give the path to generate files into", default = settings.NUMBER_DIR)
```
Please do not use `give the <real_help_message>` structure.

rerobika replied:
`help = "Output file name"`

matedabis replied:
Changed description.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+    parser.add_argument("--test-count", help = "give the number of tests to run", nargs = 1, default = [500],
+    type = int)
+    parser.add_argument("--output", help = "give the path to generate files into", default = settings.NUMBER_DIR)
+    parser.add_argument("--operand-count", help = "give the number of maximum operands in an expression",
```
Ditto.

matedabis replied:
Changed description

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+    parser.add_argument("--output", help = "give the path to generate files into", default = settings.NUMBER_DIR)
+    parser.add_argument("--operand-count", help = "give the number of maximum operands in an expression",
+    nargs = 1, default = [10], choices = range(2, 11), type = int)
+    parser.add_argument("--seed", help = "give the random seed value", nargs = 1, default = [10000],
```
Ditto.

rerobika replied:
`help = "Seed to used while generating random number"`

matedabis replied:
Created a variable and changed description.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+    return script_args
+
+filename = '''binary-ops-{NUMBER}.js'''
+MAX = (1<<53) - 1
```
Please add spaces around `<<` operator.

matedabis replied:
Solved this issue for the whole file.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+MAX = (1<<53) - 1
+MIN = -1 * MAX
+BITMAX = 9999999
+BITMIN = -9999999
```
`BITMIN = -BITMAX`

rerobika replied:
Where do these numbers come from? Please add a description for all of them.

matedabis replied:
I changed some of them and added a description for constants and variables.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+        self.operands = [0] * num_operands
+        self.operators = [0] * (num_operands -1)
+
+        if (bit == False):
```
Ditto.

matedabis replied:
Solved this issue for the whole file.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+        self.num_operands = num_operands
+        self.operators_list = []
+        for i in range(len(options.op)):
+            if (bit == False):
```
If `bit` is a Boolean value use `not bit` instead of comparing it to `False`.

matedabis replied:
Solved this issue for the whole file.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+                if (options.op[i] == "^"):
+                    self.operators_list.append(self.bitxor)
+        self.operands = [0] * num_operands
+        self.operators = [0] * (num_operands -1)
```
Use spaces around `-` operator.

matedabis replied:
Solved this issue for the whole file.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+        num_operands = options.operand_count[0]
+        self.num_operands = num_operands
+        self.operators_list = []
+        for i in range(len(options.op)):
```
Use `for op in options.op` and use `op` to access the list element instead of `options.op[i]`.

matedabis replied:
Fixed this.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+            if (bit == False):
+                if (options.op[i] == "+"):
+                    self.operators_list.append(self.add)
+                if (options.op[i] == "-"):
```
Use `elif`.

matedabis replied:
Replaced with `elif`.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+                    self.operators_list.append(self.add)
+                if (options.op[i] == "-"):
+                    self.operators_list.append(self.sub)
+                if (options.op[i] == "*"):
```
Ditto.

matedabis replied:
Replaced with `elif`.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+                    self.operators_list.append(self.sub)
+                if (options.op[i] == "*"):
+                    self.operators_list.append(self.mul)
+                if (options.op[i] == "/"):
```
Ditto.

matedabis replied:
Replaced with `elif`.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+                    self.operators_list.append(self.mul)
+                if (options.op[i] == "/"):
+                    self.operators_list.append(self.div)
+                if (options.op[i] == "%"):
```
Ditto.

matedabis replied:
Replaced with `elif`.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+        self.operators = [0] * (num_operands -1)
+
+        if (bit == False):
+            for i in range(num_operands -1):
```
Add space as above.

matedabis replied:
Solved this issue for the whole file.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+            for i in range(num_operands -1):
+                self.operands[i] = random.randint(MIN, MAX)
+                self.operators[i] = random.randint(0, len(self.operators_list) -1)
+            self.operands[num_operands -1] = random.randint(MIN, MAX)
```
Ditto.

matedabis replied:
Solved this issue for the whole file.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+            for i in range(num_operands -1):
+                self.operands[i] = random.randint(BITMIN, BITMAX)
+                self.operators[i] = random.randint(0, len(self.operators_list) -1)
+            self.operands[num_operands -1] = random.randint(BITMIN, BITMAX)
```
These 2 branches only differs in the `MIN/MAX, BITMIN/BITMAX` value. Do not copy the whole for cycle, only modify the ranges according to the condition.

matedabis replied:
Created an if statement for setting a `MAXNUM` and `MINNUM` depending on the operation type, now it has only 1 for cycle.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+        sum = op1 + op2
+        if sum < MIN:
+            op2 = MIN - random.randint(MIN, op1)
+            sum = op1 + op2
```
Remove this line.

matedabis replied:
I need this, because in this case I modify `op2` and so I have to modify `sum`

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+            sum = op1 + op2
+        elif sum > MAX:
+            op2 = MAX - random.randint(op1, MAX)
+            sum = op1 + op2
```
Ditto.

matedabis replied:
I need this, because in this case I modify `op2` and so I have to modify `sum`

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+            self.operands[num_operands -1] = random.randint(BITMIN, BITMAX)
+
+    def add(self, op1, op2):
+        sum = op1 + op2
```
Move it before the return statement (the ifs may override the `op2` value) so this addition is calculated once.

matedabis replied:
I need this for testing if it fits the interval

rerobika replied:
I see now. Okay.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+        elif sum > MAX:
+            op2 = - (MAX - random.randint(op1, MAX))
+            sum = op1 - op2
+        return {'oper': '-', 'op2': op2, 'sum': sum}
```
Same as for `add`.

matedabis replied:
I need this, because in this case I modify `op2` and so I have to modify `sum`

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+            op2 = math.floor(op2)
+        else:
+            op2 = math.ceil(op2)
+        sum = op1 * op2
```
Ditto.

matedabis replied:
I need this, because in this case I modify `op2` and so I have to modify `sum`

rerobika replied:
Okay.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+
+    def div(self, op1, op2):
+        if op2 == 0:
+            op2 = 2
```
Where does this magic number come from?

matedabis replied:
Replaced this with a random, that guaranteed not to be zero.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+        if sum > 0:
+            sum = int(math.floor(sum))
+        else :
+            sum = int(math.ceil(sum))
```
Why do we need to round this values?

matedabis replied:
Because if it might be a float, and `sum` might jump out of the interval if `op2` is not rounded.

rerobika replied:
I think `int(sum)` is more than enough, hence it "cuts" the fraction part, so the result cannot be bigger then the end of the interval.

rerobika replied:
This request does not seem resolved for me.

loki04 replied:
Did you check if the `int(sum)` is enough? What does the standard says?

rerobika replied:
_class int(x, base=10)_

Return an integer object constructed from a number or string x, or return 0 if no arguments are given. If x is a number, it can be a plain integer, a long integer, or a floating point number. **If x is floating point, the conversion truncates towards zero.** If the argument is outside the integer range, the function returns a long object instead.

This snippet is part of the python API [docs](https://docs.python.org/2/library/functions.html#int).

So it think `int(sum)` is still suitable for this case.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+
+    def mod(self, op1, op2):
+        if op2 == 0:
+            op2 = 2
```
Ditto, where does this come from?

matedabis replied:
Replaced this with a random, that guaranteed not to be zero.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+    def bitshiftleft(self, op1, op2):
+        op2 = int(math.fabs(op2))
+        if op1 < 0:
+            op2 = math.floor(math.log(MIN/op1))
```
For here and overall in this file. Use spaces around operators.

matedabis replied:
Fixed this issue for the whole file.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+            if op1 > 0:
+                op2 = math.floor(math.log(MAX/op1, 2))
+            else:
+                op2 = math.ceil(math.log(MAX/op1, 2))
```
I'm still not sure why we have to round this value. Could you explain it?

matedabis replied:
Because if it might be a float, and `sum` might jump out of the interval if `op2` is not rounded.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+        sum = op1 & op2
+        if sum < MIN:
+            op2 = sum | op1
+        elif sum > MAX:
```
remove this `elif` branch.

matedabis replied:
Are you sure that it can not  go above the maximum?

rerobika replied:
I am. Moreover. op1 and op2 are random numbers in [MIN, MAX] range. Using bitwise and operator in this values result a number in the same range, so either these check are necessary.

loki04 replied:
You can remove both checks. The bitwise operators are just set or reset bits in the same ranges.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+        sum = op2 ^ op1
+        return {'oper': '^', 'op2': op2, 'sum': sum}
+
+    def debug(self):
```
What is this function used for?

matedabis replied:
I will remove this

matedabis replied:
Removed

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+    new_file_begin = True
+    bitcalc = False
+    arithcalc = False
+    for i in range(len(options.op)):
```
Same as above. Use `for op in options.op`

matedabis replied:
Fixed

loki04 replied:
Did you really refactor this code? Use the element itself instead of the index (as I see you have never reused the index other than calculating the element)

rerobika replied:
It seems unresolved.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+    for i in range(len(options.op)):
+        if ((options.op[i] == "+") or (options.op[i] == "-") or (options.op[i] == "*") or (options.op[i] == "/") or
+        (options.op[i] == "%")):
+            arithcalc = True
```
How about self.{aritht,bit}_calc ? This can be determined while parsing the arguments.

matedabis replied:
That is good idea I will work on that.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+    for i in range(options.test_count[0]):
+        bit = False
+        counter = counter + 1
+        if (counter % 800 == 0):
```
Where does this `800` come from? Please do not use magic numbers.

matedabis replied:
Fixed

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+    file_to_open = os.path.join(options.output, filename.format(NUMBER = file_count))
+    for i in range(options.test_count[0]):
+        bit = False
+        counter = counter + 1
```
Use `+=1`

matedabis replied:
Fixed

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+        counter = counter + 1
+        if (counter % 800 == 0):
+            new_file_begin = True
+            file_count = file_count + 1
```
Use `+=1`

matedabis replied:
Fixed

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+                file.write(validate.validate)
+                new_file_begin = False
+            SEED += 1
+            if (bitcalc == True and arithcalc == False):
```
`if bitcalc and not arithcalc`

matedabis replied:
Fixed

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+                bit = True
+                gs = GenSequence(options, bit)
+                print("lol")
+            elif (bitcalc == False and arithcalc == True):
```
Use the same syntax as above.

matedabis replied:
Fixed

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,281 @@
+            else:
+                bit = not bit
+                gs = GenSequence(options, bit)
+                print("lol")
```
:) Remove this line

matedabis replied:
Removed

rerobika commented on src/scripts/templates/validate.py:
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env python
+
+validate = '''
```
This is the duplicate of #1.

rerobika reviewed (changes requested):
In general:
 - Use spaces around operators (`1+1` is not allowed use `1 + 1` instead)
 - Add a documentation for each function, constants
 - Remove magic numbers.

matedabis committed:
Add binary operator tests

Added operator tests for + - % / * arithmetic operators, and | ^ & << >> >>> bitwise operators.

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+import argparse
+import settings
+import collections
+path = os.path.join(settings.TEMPLATE_DIR)
```
Leave a comment here that you are looking for local packages.
And move this local import to the end of imports

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+import math
+import subprocess
+
+BINARY_OPERATORS = ["+", "-", "div", "*", "%", "^", "&", "|", ">>", "<<"] # The binary operators used in this script
```
I guess the followings are hidden variables (and/or constants). So, I suggest to user the underline '_' prefix in the variables names.

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+TEST_CASE_IN_A_FILE = 800 # The number of test cases before start new file
+TEST_RUNNER_SCRIPT = os.path.join(settings.SCRIPTS_DIR, 'runscripts.sh')
+
+def get_arguments():
```
Since it will be only used if this script is the main one. So, make it hidden or move the content inside the main function.

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+def get_arguments():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--op", help = "List of binary operator(s) from an expression is being generated", nargs = '+',
+    default = BINARY_OPERATORS, choices = BINARY_OPERATORS)
```
BINARY_OPERATORS is used only here. Create a local variable for this.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+DEFAULT_SEED = 10000 # The default value of random seed
+GENERATED_FILENAME = '''binary-ops-{NUMBER}.js''' # The file name prefix of the generated tests
+MAX = (1 << 53) - 1 # The max value of random number
+MIN = -1 * MAX  # The min value of random number
```
This is strange for me. If we using power of 2 limits then:

```
MAX = (1 << 53) - 1 
```

rerobika replied:
This is strange for me. If we using power of 2 limits then:

```
MAX = (1 << 53) - 1 
MIN = -(1 << 53)
```

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+MAX = (1 << 53) - 1 # The max value of random number
+MIN = -1 * MAX  # The min value of random number
+BITMAX = 1 << 32 - 1 # The max value of random number for bitwise operations
+BITMIN = - BITMAX # The min value of random number for bitwise operations
```
Ditto.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+BITMIN = - BITMAX # The min value of random number for bitwise operations
+BITMAX_EXPOSANT = 31 # Maximum number for safely shifting an integer
+BITMIN_EXPOSANT = 0  # Because of bit operations are pecise only in 32 bit
+TEST_CASE_IN_A_FILE = 800 # The number of test cases before start new file
```
Constants are now okay, but can we separate it to an other file? e.g `constans.py`?

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+BITMAX_EXPOSANT = 31 # Maximum number for safely shifting an integer
+BITMIN_EXPOSANT = 0  # Because of bit operations are pecise only in 32 bit
+TEST_CASE_IN_A_FILE = 800 # The number of test cases before start new file
+TEST_RUNNER_SCRIPT = os.path.join(settings.SCRIPTS_DIR, 'runscripts.sh')
```
This variable it not used.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+
+    script_args = parser.parse_args()
+
+    return script_args
```
I'd also separate it to an other script.

rerobika replied:
Still I'd separate his arg parse part. See #6 .

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+                    arithcount += 1
+                elif (op == "%"):
+                    self.operators_list.append(self.mod)
+                    arithcount += 1
```
How about?

```py
if (not bit):
                if (op == "+"):
                    self.operators_list.append(self.add)
                elif (op == "-"):
                    self.operators_list.append(self.sub)
                elif (op == "*"):
                    self.operators_list.append(self.mul)
                elif (op == "div"):
                    self.operators_list.append(self.div)
                elif (op == "%"):
                    self.operators_list.append(self.mod)
                else:
                    continue # (and) print a warning message
                arithcount += 1
```

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+                    bitcount += 1
+                elif (op == ">>"):
+                    self.operators_list.append(self.bitshiftright)
+                    bitcount += 1
```
Same as above.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+        if (bit): # Setting minimum and maximum according to bitwise or arithmetic calculations
+            MAXNUM = BITMAX
+            MINNUM = BITMIN
+        else:
```
this `else` is unnecessary.

loki04 replied:
No need for this whole `else` part, since you already set the values.

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+
+class GenSequence:
+    def __init__(self, options, bit):
+        bitcount = 0
```
I don't see why it is needed.

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+        self.operators_list = []
+        for op in options.op:
+            if (not bit):
+                if (op == "+"):
```
you could use the following construct:
```
self.operators_list.append( { "+" : self.add,   "-" : self.sub,  ...}[op] );
arithcount += 1;
```

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+        # Generating the expression
+        for i in range(num_operands - 1):
+            self.operands[i] = random.randint(MINNUM, MAXNUM)
+            if (arithcount != 1):
```
There is no need for `arithcount` use `not bit and len(options.op)` instead. In addition I guess the `len(options.op)` can be avoid as well.

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+
+    # Bitshifting to left
+    def bitshiftleft(self, op1, op2):
+        BITSHIFT_LEFTMAX = (2 ** 31) -1 # JS does not calculate precisely with numbers out of this interval
```
You used the shift operator (`(1 << 53) - 1`) in MAX, please use the same here as well (`(1 << 31) - 1`)

rerobika replied:
Isn't this `BITMAX`?

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+    def bitor(self, op1, op2):
+        sum = op1 | op2
+        if sum < MIN:
+            print("entered less\n")
```
Opps! Remove this.

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+    # Bitwise xor (exclusive or)
+    def bitxor(self, op1, op2):
+        sum = op1 ^ op2
+        if sum < BITMIN:
```
None of these bitwise checks are needed. Same is true for the others.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,347 @@
+        or options.op[i] == "<<"):
+            bitcalc = True
+
+    if (bitcalc == True and arithcalc == False):
```
Do not use `boolean_var == {True, False}`

matedabis commented:
Thanks for the lot of reviews and the time spent reviewing my patch, they really helped improvement. Excluding the reviews mentioned below I agreed and corrected the mistakes made in the code. The script for constants is committed to another PR.
https://github.com/szeged/js-test-generators/pull/7

@rerobika 
JS max min safe integer evidence link.
 https://stackoverflow.com/questions/307179/what-is-javascripts-highest-integer-value-that-a-number-can-go-to-without-losin

BITMAX is not equal to BITSHIFT_LEFTMAX and BITMIN is not equal to BITSHIFT_LEFTMIN because of JS is working precisely on a larger interval with other bitwise operators, but left bitshifting only works precisely in a smaller interval (1 << 31 is the first number from zero calculated incorrectly).

@loki04 
I also thought that it can not run out of the interval, but it did, and I need those checks. For example assume if the interval is [-5,5] and JS does the following.
-5 & -4 == -8
 
self.operators_list.append( { "+" : self.add,   "-" : self.sub,  ...}[op] );
The problem with this is that in many cases I need that else branch, so I cannot use this, because in each expression I need to pick all the added operators of a type (bitwise or arithmetic) and for example if I need to pick bitwise operators, I go trough the operators, and if I reach "+" or "-" I have to skip it, and this construct does not allow me to add a default case to continue.

@repasics is working on putting the parser in a script, so I will fix the function get_arguments() according to that script as soon as he commits his work.

loki04 commented:
> BITMAX is not equal to BITSHIFT_LEFTMAX and BITMIN is not equal to BITSHIFT_LEFTMIN because of JS is working precisely on a larger interval with other bitwise operators, but left bitshifting only works precisely in a smaller interval (1 << 31 is the first number from zero calculated incorrectly).

There is no incorrect calculation! Please, check out the standard ( http://www.ecma-international.org/ecma-262/6.0/#sec-left-shift-operator-runtime-semantics-evaluation )

Lets say you have the following JS snipets:
```
> ((1 << 30)).toString(16)
'40000000'
> ((1 << 31)).toString(16)
'-80000000'
> ((1 << 32)).toString(16)
'1'
> ((1 << 33)).toString(16)
'2'
> ((1 << 34)).toString(16)
'4'
> ((1 << 63)).toString(16)
'-80000000'
```
You have to emulate this with python. For example:
```
In [41]: hex((1 << (30 & 0x1f)))
Out[41]: '0x40000000'
In [42]: hex((1 << (31 & 0x1f)))
Out[42]: '0x80000000'
In [43]: hex((1 << (32 & 0x1f)))
Out[43]: '0x1'
In [44]: hex((1 << (33 & 0x1f)))
Out[44]: '0x2'
In [45]: hex((1 << (34 & 0x1f)))
Out[45]: '0x4'
In [46]: hex((1 << (63 & 0x1f)))
Out[46]: '0x80000000'
```

> I also thought that it can not run out of the interval, but it did, and I need those checks. For example assume if the interval is [-5,5] and JS does the following.
> -5 & -4 == -8

That is not true. The range is 32 bits (int32). None of the bitwise operands will jump over that barrier!
See:
```
> (-((1 << 31) -1) & -((1 << 31) -3)).toString(16)
'-7fffffff'
```

In other words, if you check your example ...
```
> ((-4) >>> 0).toString(16)
'fffffffc'
> ((-5) >>> 0).toString(16)
'fffffffb'
> ((-4 & -5) >>> 0).toString(16)
'fffffff8'
```

matedabis commented:
Fixed according to @loki04's suggestion.

matedabis commented:
Modified according to reviews in this PR:
https://github.com/szeged/js-test-generators/pull/7
Now this script is able to test bitshifting including all the numbers in the 32 bit interval.

Should I also change `MAXNUM`, `MINNUM` and `SEED` to low case, or can they stay like this, since they are not part of the constants?

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,300 @@
+                    self.operators_list.append(self.bitor)
+                elif (op == "shiftleft"):
+                    self.operators_list.append(self.bitshiftleft)
+                elif (op == "shiftright"):
```
How about `unsigned right shift`? This operator is also available in JS.

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,300 @@
+#!/usr/bin/env python
+
+
```
Use only one empty line here.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,300 @@
+        sum = int(hex((op1 >> ((op2 % (_constants.bitmax_exposant + 1)) & 0x1f))), 16)
+        return {'oper': '>>', 'op2': op2, 'sum': sum}
+
+    # Bitwise addition
```
`Bitwise addition` ?

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,300 @@
+        os.makedirs(options.output)
+    SEED = options.seed
+    random.seed(SEED)
+    counter = 0 # How many tests are written (defending against script error: null)
```
`How many tests are written (defending against script error: null)` -> 'Generated file count'

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,300 @@
+    if (bitcalc and not arithcalc):
+        bit = True
+
+    print('\033[94m\nGenerating tests\n\033[0m')
```
`\033[94m` also part of #5.

matedabis commented:
Corrected according to the reviews, expected at the checking of `arithcalc` and `bitcalc`.
Now the script uses the argparser from PR https://github.com/szeged/js-test-generators/pull/6
Proof for the `unsigned32` conversation at the unsigned right shift:
https://stackoverflow.com/a/11418436

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,286 @@
+        op2 = random.randint(_constants.bitmin, _constants.bitmax)
+        if (op1 >= 0): # For positive operand 1 it works the same as bitshiftright
+            sum = int(hex((op1 >> ((op2 % (_constants.bitmax_exposant + 1)) & 0x1f))), 16)
+        else: # For negative operand 1 it needs conversion to unsigned
```
Still has issue related to the standard. See the Runtime Semantics: Evaluation part at https://www.ecma-international.org/ecma-262/6.0/#sec-unsigned-right-shift-operator . The op1 and op2 have to be unsigned int32.

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,286 @@
+
+        # If bitwise operators are currently needed
+        elif ((op == "xor") or (op == "and") or (op == "or") or (op == "shiftright")
+        or (op == "shiftleft") or (op == "shiftrightzero")):
```
Fix the indentation.

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,286 @@
+            os.remove(file_to_open)
+        with open(file_to_open, "a") as file: # If new file begins validate function needs to be added
+            if (new_file_begin):
+                print(file_to_open)
```
Is there a need for this printing?

matedabis replied:
@repasics also included it in his script, and this is the only way to know which files we generated. (If we generate 5 tests and run it, and its correct, and we still get an error, it can be because of a previous generation of a higher test count) and this way we know it it could be because of that. And if we are generating like 50.000+ tests, it is good to see where we are in the generating. And the test runner script does not shows us this. I think we need it here and it is useful.
https://github.com/szeged/js-test-generators/blob/69e7b7f2faffd704d139bac1bcf509dfc24d2c29/src/scripts/generate-unary-ops.py#L39

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,286 @@
+                    self.operators_list.append(self.bitshiftleft)
+                elif (op == "shiftright"):
+                    self.operators_list.append(self.bitshiftright)
+                elif (op == "shiftrightzero"):
```
`shiftrightzero`? It's a bit strange alias for `<<<`.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,286 @@
+    # Division (floor division)
+    def div(self, op1, op2):
+        if op2 == 0: # To avoid division by zero
+            if (random.randint(0,1)):
```
spaces around `,`

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,286 @@
+                op2 = random.randint(1, _constants.max)
+            else:
+                op2 = random.randint(_constants.min, -1)
+        if (((op1 < 0 and op2 > 0) or (op1 > 0 and op2 < 0)) and (op1 % op2 != 0)):
```
`((op1 < 0 and op2 > 0) or (op1 > 0 and op2 < 0))`

This check means that `op1` and `op2` signs differ, so
`op1 * op2 < 0` results the same ( `+ * - = -`,  `- * + = -`)

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,286 @@
+    bit = False # Actually are we doing bitwise or arithmetic calculations
+    for op in options.op:
+        # If arithmetic operators are currently needed
+        if ((op == "add") or (op == "sub") or (op == "mul") or (op == "div") or
```
``` py
if op in ["add", "sub", "mul", "div"]
```

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,286 @@
+            arithcalc = True
+
+        # If bitwise operators are currently needed
+        elif ((op == "xor") or (op == "and") or (op == "or") or (op == "shiftright")
```
ditto

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,286 @@
+        if (generated_test_counter % _constants.test_case_in_a_file == 0): # If a new file is needed
+            new_file_begin = True
+            file_count += 1
+            file_to_open = os.path.join(options.output, _GENERATED_FILENAME.format(NUMBER = file_count))
```
How about the following concept: (#3 is might related as well)

Each {unary_binary} generator script **ONLY** generates test cases, and passes it's output to a specific _file writer_ script? (which cuts, formats, separates, ... the input according to what is needed.)

matedabis commented:
Updated according to the change requests and discussions. `_write_file` module is not yet completed by @repasics , but it will be needed to call this way.
Important changes: 

- It will use a file writer module instead of directly writing the file by the script (`_write_file`)

- I needed to add 2 new constants, because of unsigned right shift works on a different interval than arithmetic operations and all the other bitwise operations, and because of that, I have to handle it as a 3rd case of expression type.

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,312 @@
+
+_GENERATED_FILENAME = '''binary-ops-{NUMBER}.js''' # The file name prefix of the generated tests
+_BINARY_OPERATORS = ["add", "sub", "div", "mul", "mod", "xor", "and", "or", "rightshift", "leftshift",
+"unsignedrightshift"] # The binary operators used in this script
```
How about to add these to _constants? For example:
```
import const
### const.py:
# class _constants:
#  a = 42

class _constants(const._constants):
  b = 10

print(_constants.a, _constants.b)
```

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,312 @@
+"unsignedrightshift"] # The binary operators used in this script
+
+# Needed for unsigned right shift
+def unsigned32(signed):
```
Until there is no common util.py for these kind of functions, you should pack them to the main class.

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,312 @@
+                    self.operators_list.append(self.div)
+                elif (op == "mod"):
+                    self.operators_list.append(self.mod)
+                else:
```
You could use the following structure (or a similar one), and simplify not just this, but other code sequences below:
```
class A:
  def __init__(self):
    self.oper_func = {
      "add": self.add,
      "sub": self.sub,
      "mod": self.mod
    }

  def add(self):
    print("Function: add")
  def sub(self):
    print("Function: sub")
  def mod(self):
    print("Function: mod")

  def test(self, text):
    if text in self.oper_func:
      self.oper_func[text]()
    else:
      print("Not found")

a = A()

for t in ["add", "mod", "nothing", True]:
  a.test(t)
```

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,312 @@
+        if (calc == 2):
+            MAXNUM = _constants.bitmax
+            MINNUM = _constants.bitmin
+        elif(calc == 3):
```
Space between `elif` and ` calc == 3`.

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,312 @@
+        # Setting max and min according to the calculations
+        MAXNUM = _constants.max
+        MINNUM = _constants.min
+        if (calc == 2):
```
Do not use parenthesis here.

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,312 @@
+    # Bitwise operations
+
+    # Bitshifting to left
+    def bitleftshift(self, op1, op2):
```
Please use the same steps, or at least the same logic to evaluate the operator as the standard says ( http://www.ecma-international.org/ecma-262/6.0/#sec-left-shift-operator-runtime-semantics-evaluation ).

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,312 @@
+        return {'oper': '^', 'op2': op2, 'sum': sum}
+
+# Generator function
+def generate_binary_tests(options):
```
I was wondering why you do not put this function inside the class? And refactor it to smaller parts? I think you should give it a try, probably the readability will be improved with this.

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,312 @@
+            file_content += ("validate_numeric((%s), (%s), (%s));\n" % (expression, str(expected_value), false_values))
+        write_file._write_file(options.output,  _GENERATED_FILENAME.format(NUMBER = file_count), file_content)
+
+def main(options):
```
I still do not like the this kind of main functions. It is so simple that you could rename the `generate_binary_tests` to `main` instead! Or if you refactor the function as I said above, it will have some valuable actions here as well.

rerobika commented on src/scripts/constants.py:
```diff
@@ -24,5 +24,7 @@ class _constants:
     min = - 1 * max             # The min value of random number (See also Number.MIN_SAFE_INTEGER)
     bitmax = (1 << 31) - 1      # The max value of random number for bitwise operations
     bitmin = - bitmax - 1       # The min value of random number for bitwise operations
+    uint32max = (1 << 32) -1    # The max value of random number for unsigned right shift (See also unsigned int 32)
```
spaces around `-`.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,318 @@
+from util import unsigned32
+
+_GENERATED_FILENAME = '''binary-ops-{NUMBER}.js''' # The file name prefix of the generated tests
+_BINARY_OPERATORS = ["add", "sub", "div", "mul", "mod", "xor", "and", "or", "rightshift", "leftshift",
```
Isn't `_BINARY_OPERATORS` == `_ARITHMETIC_OPERATORS` + `_BITWISE_OPERATORS` + `_UINT32_OPERATORS`. If it is true, then do not duplicate the elements.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,318 @@
+_BITWISE_OPERATORS = ["xor", "and", "or", "rightshift", "leftshift"]  # The bitwise operators used in this script
+_UINT32_OPERATORS = ["unsignedrightshift"]
+
+class GenSequence:
```
Please add description for the class and for each member.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,318 @@
+        }
+
+        for op in options.test_cases:
+            if calc == 1:
```
Cannot we use symbolic constants for `calc`? `1,2,3` does not provide too detailed information about what it really does.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,318 @@
+        self.operands = [0] * num_operands # Creating a list for operands according to number of operands
+        self.operators = [0] * (num_operands - 1) # Creating a list for operators according to number of operands
+
+        # 1: arithmetic, 2: bitwise (excluding unsigned right shift) 3: unsigned right shift
```
See the comment above.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,318 @@
+            if op1 != 0:
+                op2 = (_constants.bitmax_exposant - 1) - int.bit_length(op1)
+                if op2 < 0: # Operand 2 also cant go below zero
+                    op2 = 0
```
`op2 = max (op2, 0)`

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,318 @@
+
+    for op in options.test_cases:
+        # If arithmetic operators are currently needed
+        if op in ["add", "sub", "mul", "div", "mod"]:
```
This list has already declared previously, see `_ARITHMETIC_OPERATORS`.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,318 @@
+            arithcalc = True
+
+        # If bitwise operators are currently needed
+        if op in ["xor", "and", "or", "rightshift", "leftshift"]:
```
Ditto.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,318 @@
+            bitcalc = True
+
+        # If unsigned right shift is currently needed
+        if op in ["unsignedrightshift"]:
```
Ditto.

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,318 @@
+        calc = (1 + random.randint(0, 1) * 2) # Must be 1 or 3
+    elif arithcalc and bitcalc and urscalc:
+            calc = random.randint(1, 3) # 1, 2 or 3
+    return calc
```
Same as above.... Use symbolic constants!

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,318 @@
+            MINNUM = _constants.bitmin
+        elif calc == 3:
+            MAXNUM = _constants.uint32max
+            MINNUM = _constants.uint32min
```
Cannot we use `set_min_max` here?

rerobika commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,318 @@
+                # or shift right zero (3) - No (0) is unsetted
+    arithcalc = needed_calculation_types(options)["arithcalc"]
+    bitcalc = needed_calculation_types(options)["bitcalc"]
+    urscalc = needed_calculation_types(options)["urscalc"]
```
After this point, the following shouldn't the `file_writer` script's task? (except to generate the content)

matedabis replied:
We respectfully disagree with this, because we use different methods for everything excepted the file writing. But I corrected and agreed with all the other reviews of yours.

rerobika replied:
So you mean, you really need this whole file{writing, cutting, separating} part in every generator script??? The only difference between each script is the content. By the `file_writer` script I meant an universal script which gets a `relatively big {string, list, dictionary}` whatever is needed (thats what the generator scripts produce) and **THIS** script do all the job, what is needed to create the file(s) (cutting, separating).

loki04 commented on src/scripts/generate-binary-ops.py:
```diff
@@ -0,0 +1,317 @@
+            if op1 != 0:
+                op2 = (_constants.bitmax_exposant - 1) - int.bit_length(op1)
+                op2 = max(op2, 0) # Operand 2 also cant go below zero
+        rnum = unsigned32(op2)
```
You do this twice in case all the above conditions are false. Why?

matedabis commented:
This PR and #10 together are working and up to date to each other.

This way generated files look like this:
`
function validate_numeric(test_value, expected_result, false_results) {
    var radix = 2 + Math.abs(expected_result % 35);
    assert(test_value === expected_result);
    assert(test_value.toString(radix) === expected_result.toString(radix));
    for (var i in false_results) {
        radix = 2 + Math.abs(false_results[i] % 35);
        assert(test_value !== false_results[i]);
        assert(test_value.toString(radix) !== false_results[i].toString(radix));
    }
}

t1 = 1393823298;
t2 = 1773334627;
t3 = 0;
t4 = -1624120051;
t5 = 1273819165;
t6 = -799240456;
t7 = 84806027;
t8 = 1521124651;
t9 = -1464979674;
t10 = 2124382107;
f1 = 1026581450;
f2 = 2130015057;
f3 = 2134190604;
f4 = 255241881;
f5 = 2133467548;
validate_numeric((((((((((t1 & t2) << t3) & t4) & t5) & t6) >> t7) & t8) & t9) | t10)
, 2124382107, [f1, f2, f3, f4, f5]);

...
`

matedabis commented:
This PR and #10 together are working and up to date to each other.

This way generated files look like this:
```
function validate_numeric(test_value, expected_result, false_results) {
var radix = 2 + Math.abs(expected_result % 35);
assert(test_value === expected_result);
assert(test_value.toString(radix) === expected_result.toString(radix));
for (var i in false_results) {
radix = 2 + Math.abs(false_results[i] % 35);
assert(test_value !== false_results[i]);
assert(test_value.toString(radix) !== false_results[i].toString(radix));
}
}

t1 = 1393823298;
t2 = 1773334627;
t3 = 0;
t4 = -1624120051;
t5 = 1273819165;
t6 = -799240456;
t7 = 84806027;
t8 = 1521124651;
t9 = -1464979674;
t10 = 2124382107;
f1 = 1026581450;
f2 = 2130015057;
f3 = 2134190604;
f4 = 255241881;
f5 = 2133467548;
validate_numeric((((((((((t1 & t2) << t3) & t4) & t5) & t6) >> t7) & t8) & t9) | t10)
, 2124382107, [f1, f2, f3, f4, f5]);

...
```

rerobika commented on src/scripts/constants.py:
```diff
@@ -24,5 +24,7 @@ class _constants:
     min = - 1 * max             # The min value of random number (See also Number.MIN_SAFE_INTEGER)
     bitmax = (1 << 31) - 1      # The max value of random number for bitwise operations
     bitmin = - bitmax - 1       # The min value of random number for bitwise operations
+    uint32max = (1 << 32) - 1    # The max value of random number for unsigned right shift (See also unsigned int 32)
```
The comment is misaligned.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,325 @@
+_ARITHMETIC_OPERATORS = ["add", "sub", "div", "mul", "mod"]  # The arithmetic operators used in this script
+_BITWISE_OPERATORS = ["xor", "and", "or", "rightshift", "leftshift"]  # The bitwise operators used in this script
+_UINT32_OPERATORS = ["unsignedrightshift"]
+_BINARY_OPERATORS = _ARITHMETIC_OPERATORS + _BITWISE_OPERATORS + _UINT32_OPERATORS # The binary operators
```
Values are okay now, just add it to a local class.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,325 @@
+# 1: arithmetic, 2: bitwise (excluding unsigned right shift) 3: unsigned right shift
+_ARITH = 1
+_BIT = 2
+_UINT32 = 3
```
Ditto.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,325 @@
+            calc = random.randint(2, 3) # 2 or 3 (_BIT or _UINT32)
+        elif arithcalc and urscalc and not bitcalc:
+            calc = (1 + random.randint(0, 1) * 2) # Must be 1 or 3 (_ARITH or _UINT32)
+        elif arithcalc and bitcalc and urscalc:
```
This if/elif crowd is ridiculous. Could you please try to simplify it?

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,325 @@
+        for op in self.options.test_cases:
+            # If arithmetic operators are currently needed
+            if op in _ARITHMETIC_OPERATORS:
+                arithcalc = True
```
`arithcalc =op in _ARITHMETIC_OPERATORS `

matedabis replied:
I see your point here, but I can not simplify these 2 lines to 1 line, because if it is false, I must not change the value, or do you mean to change only the last line to this suggestion? (The same question for the other 2 below)

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,325 @@
+                arithcalc = True
+
+            # If bitwise operators are currently needed
+            if op in _BITWISE_OPERATORS:
```
Ditto.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,325 @@
+                bitcalc = True
+
+            # If unsigned right shift is currently needed
+            if op in _UINT32_OPERATORS:
```
Same as above.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,325 @@
+        return {"arithcalc" : arithcalc, "bitcalc" : bitcalc, "urscalc" : urscalc}
+
+    # Generator
+    def generate_binary_ops(self, generated_filename ):
```
unnecessary space before `)`

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,325 @@
+        bitcalc = self.needed_calculation_types()["bitcalc"]  # If we need to do bitwise calculations
+        urscalc = self.needed_calculation_types()["urscalc"] # If we need to do unsigned right shift
+                                                               # (abbreviation: unsigned right shift calculation)
+        calc = 0 # Actually are we doing arithmetic (1) or bitwise (excluding unsigned right shift) (2) calculations
```
`calc` value is overwritten in each iteration below, since this declaration is unnecessary.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,325 @@
+        calc = 0 # Actually are we doing arithmetic (1) or bitwise (excluding unsigned right shift) (2) calculations
+                 # or shift right zero (3) - No (0) is unsetted
+        test_case_in_a_file = _constants.test_case_in_a_file
+        # Over 8 operands jerry can not handle the original number of test cases
```
This generator is not specified for `JerryScript`!!!!

rerobika reviewed (changes requested):
Getting closer!

matedabis commented:
Updated according to the review, #11 and #10 did not need modifications because of these corrections, and they work correctly together.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,305 @@
+    _BINARY_OPERATORS = _ARITHMETIC_OPERATORS + _BITWISE_OPERATORS + _UINT32_OPERATORS # The binary operators
+    # used in this script
+    _NEEDED_CALCULATIONS = [] # Actually needed calculation types represented by their sybolic constants
+    _TYPES_COUNT = 0 # Number of calculation types needed
```
Use local class for storing these constants.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,305 @@
+# Class for expressions
+class GenSequence(Generator):
+
+    _GENERATED_FILENAME = '''binary-ops-{NUMBER}.js''' # The file name prefix of the generated tests
```
This constant will be passed to the file_write, in a setting object. How about declaring the settings object here?

matedabis replied:
It is not possible, because there is no way to see `self.options.output` from the inner class, because it is in the outer class.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,305 @@
+
+    # Generating the expression with operands and operators
+    def expression_generator(self, calc):
+        self.operators_list = [] # List of currencly used operators
```
Typo: currently

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,305 @@
+        return {'oper': '^', 'op2': op2, 'sum': sum}
+
+    # Setting maximum and minimum depending on operation type
+    # 1: arithmetic, 2: bitwise (excluding unsigned right shift) 3: unsigned right shift
```
The comment seems outdated after the symbolic constants have been introduced.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,305 @@
+    def string_creator(self, MINNUM, MAXNUM):
+        last = self.operands[0]
+        expression = ""
+        expression += ("%s%s" % ("(" * (self.num_operands - 1), self.TEST_VALUE_VAR_NAME.format(NUMBER = 1)))
```
Cannot we assing it directly to `expression`?

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,305 @@
+        expression += ("%s%s" % ("(" * (self.num_operands - 1), self.TEST_VALUE_VAR_NAME.format(NUMBER = 1)))
+        for i in range(self.options.operand_count - 1):
+            result = self.operators_list[self.operators[i]](last, self.operands[i + 1])
+            expression += (" %s %s)" % (result["oper"], self.TEST_VALUE_VAR_NAME.format(NUMBER = (i + 2))))
```
The additional whitespace character is needed before the first %s?

matedabis replied:
It is needed, because this way the expression looks like
```
(((((((((t1 | t2) >> t3) | t4) & t5) >> t6) | t7) & t8) | t9) << t10)
```
Without that there wouldn't be spaces before the operators.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,305 @@
+
+    # Calculating currently needed calculation types
+    def needed_calculation_types(self):
+
```
Unnecessary new line

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,305 @@
+        self._TEST_CASE_IN_A_FILE)
+        for i in range(self.options.test_count): # In range of number test cases
+            # Actually are we doing arithmetic or bitwise (excluding unsigned right shift) calculations
+            # or shift right zero
```
Shift right zero?

matedabis commented:
Updated according to the review and also updated #11 according to this generator.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,317 @@
+    # Setting maximum and minimum depending on operation type
+    def set_min_max(self, calc):
+        if calc == self._InnerValues.arith:
+            MINNUM = _constants.min
```
make it lowercase

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,317 @@
+            MAXNUM = _constants.bitmax
+        else:
+            MINNUM = _constants.uint32min
+            MAXNUM = _constants.uint32max
```
Same as above.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,317 @@
+        return {'test_case' : test_case, 'false_numbers_string' : false_numbers_string}
+
+    # Converting expression, expected value, false value in a function call to string format
+    def string_creator(self, MINNUM, MAXNUM):
```
How about using the result object of `set_min_max` function instead of using it's properties separately?

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,317 @@
+                expression = "(" + expression
+            self.operands[i + 1] = result["op2"]
+            last = result["sum"]
+        expression += "\n"
```
After this point the code look really similar to #11 ending part. It'd be great to create a function for it to reduce the code duplication.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,317 @@
+            # or unsigned right shift
+            calc = self.actual_calc()
+            self.expression_generator(calc)
+            self.test_case += self.string_creator(self.set_min_max(calc)["MIN"], self.set_min_max(calc)["MAX"])
```
See the comment above.

matedabis replied:
#11 uses exactly this function

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,320 @@
+
+        # Setting max and min according to the calculations
+        maxnum = self.set_min_max(calc)["max"]
+        minnum = self.set_min_max(calc)["min"]
```
We should only call the `self.set_min_max(calc)` only once.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,320 @@
+        elif sum > _constants.max:
+            op2 = _constants.max - random.randint(op1, _constants.max)
+            sum = op1 + op2
+        return {'oper': '+', 'op2': op2, 'sum': sum}
```
How about creating a new function like:

``` py
def create_operation_result_object (oper, op2, sum):
  return {'oper': oper 'op2': op2, 'sum': sum}
```

After that each arithmetic/binary operations could return with:

``` py
return create_operation_result_object ('+,-,*,/, etc.', op2, sum):
```

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,320 @@
+    # Arithmetic operations
+
+    # Addition
+    def add(self, op1, op2):
```
How about renaming `op1`, `op2` and `sum`?

Since we are using binary operators, it'd be better to call the operands `left_value` and `right_value`.
Furthermore, `sum` is okay for addition, but not for subtraction, multiplication, etc... I'd also rename it to `result`.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,323 @@
+        gen_settings = settings.file_write_settings(generated_filename, self.options.output, validate.validate_numeric,
+        self._InnerValues.test_case_in_a_file)
+        for i in range(self.options.test_count): # In range of number test cases
+            # Actually are we doing arithmetic or bitwise (excluding unsigned right shift) calculations
```
we are doing ...

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,323 @@
+        bit = 2
+        uint32 = 3
+
+        needed_calculations = [] # Actually needed calculation types represented by their sybolic constants
```
Typo: sybolic

rerobika replied:
Also, can we rename this array? The 'needed' is a bit strange for me

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,323 @@
+        # Creating and adding declarations of false values
+        test_case += self.create_declarations_string(self.create_declarations_dict(false_numbers,
+        self.FALSE_RESULT_VAR_NAME))
+        # Creating and adding declarations of expected value
```
The creating and adding ... sentence is appears repeatedly. Please simplify it.

rerobika commented on src/scripts/generate_binary_ops.py:
```diff
@@ -0,0 +1,323 @@
+        declarations = self.create_declarations(expected_value, min_max)
+        test_case = declarations['test_case']
+        test_case += calculation
+        # Creating and adding the function call
```
Ditto.

loki04 reviewed (approved):
LGTM

rerobika reviewed (approved):
LGTM
